### ISSUE TITLE
enhance indexing and arithmetic safety

### ISSUE THREAD

**Issue by ralexstokes** [ralexstokes/ssz-rs#22]
analogues of these:

https://github.com/ralexstokes/ethereum-consensus/issues/151

https://github.com/ralexstokes/ethereum-consensus/issues/152

for this repo :)

**Comment by claravanstaden** [comment]
Hey @ralexstokes! As part of our production readiness of https://github.com/Snowfork/snowbridge I will be taking a look at these.

**Comment by ralexstokes** [comment]
@0xTylerHolmes found a panic here :)

```rust
use ssz_rs::prelude::*;

type Bar = Vector<u8, 96>;

#[derive(PartialEq, Eq, Debug, Default, Clone, SimpleSerialize)]
pub struct Foo {
    pub bar: Bar,
}

fn main() {
    let data = vec![];
    let foo = Foo::deserialize(&data);
    dbg!(foo);
}
```

and what fails is this: https://github.com/ralexstokes/ssz-rs/blob/main/ssz-rs-derive/src/lib.rs#L202 where the range on `encoding` hits out of bounds

instead, we should use https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get

**Comment by ralexstokes** [comment]
this should be addressed by all of the work linked here once #57 is merged

great work everyone and thank you all SO much for improving this library
